cmd/api/src/database/sso.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2024 Specter Ops, Inc.
+//
+// Licensed under the Apache License, Version 2.0
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package database
+
+import (
+	"context"
+
+	"github.com/specterops/bloodhound/src/model"
+	"gorm.io/gorm"
+)
+
+const (
+	ssoProviderTableName = "sso_providers"
+)
+
+// SSOProviderData defines the methods required to interact with the sso_providers table
+type SSOProviderData interface {
+	CreateSSOProvider(ctx context.Context, name, slug string, authType model.SessionAuthProvider) (model.SSOProvider, error)
+	CreateSSOProviderWithTransaction(ctx context.Context, tx *gorm.DB, name, slug string, authType model.SessionAuthProvider) (model.SSOProvider, error)
+}
+
+// CreateSSOProvider creates an entry in the sso_providers table
+func (s *BloodhoundDB) CreateSSOProvider(ctx context.Context, name, slug string, authType model.SessionAuthProvider) (model.SSOProvider, error) {
+	provider := model.SSOProvider{
+		Name: name,
+		Slug: slug,
+		Type: authType,
+	}
+
+	return provider, CheckError(s.db.WithContext(ctx).Table(ssoProviderTableName).Create(&provider))
+}
+
+// CreateSSOProviderWithTransaction uses a db transaction to create an entry in the sso_providers table
+func (s *BloodhoundDB) CreateSSOProviderWithTransaction(ctx context.Context, tx *gorm.DB, name, slug string, authType model.SessionAuthProvider) (model.SSOProvider, error) {
```

Comment:
We don't do this elsewhere in the codebase, but I thought it would be helpful to move the responsibility of the transaction to the associated table's db interactor.

This allows us to centralize the database interaction into a single file rather than creating bespoke transactions in other table's db interactions.

Reply:
This is fine. I want to say you can also do something like this if you want to avoid having an additional API to maintain:

``` go
	err := s.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
	    bhdb := NewBloodhoundDB(tx, s.idResolver)
	    
		if ssoProvider, err := bhdb.CreateSSOProvider(ctx, name, slug, model.SessionAuthProviderOIDC); err != nil {
			return err
		} else {
		    ...
		}
	})
```

Reply:
Ooh that's great! I'll get that in